File: src/Domain/Service/OutageFinder.php
<?php

declare(strict_types=1);

namespace App\Domain\Service;

use App\Domain\Entity\Outage;
use App\Domain\Entity\User;
use App\Domain\ValueObject\OutageInfo;

final class OutageFinder
{
    /**
     * @param Outage[] $allOutages
     */
    public function findOutageForNotification(User $user, array $allOutages): ?Outage
    {
        foreach ($allOutages as $outage) {
            if (!$outage->affectsUserAddress($user->address)) {
                continue;
            }

            $outageInfo = new OutageInfo($outage->period, $outage->description);

            if ($user->isAlreadyNotifiedAbout($outageInfo)) {
                return null;
            }

            return $outage; // Found first matching outage
        }

        return null;
    }
}

File: internal/domain/outage_finder.go
package domain

// FindOutageForNotification finds the first matching outage for a user that they haven't been notified about.
func FindOutageForNotification(user *User, allOutages []*Outage) *Outage {
	for _, outage := range allOutages {
		if !outage.AffectsUserAddress(user.Address) {
			continue
		}

		outageInfo := NewOutageInfo(outage.Period, outage.Description)

		if user.IsAlreadyNotifiedAbout(outageInfo) {
			return nil
		}

		return outage
	}

	return nil
}
